pkg/controller/core.oam.dev/v1alpha2/application/revision.go

```diff
@@ -80,9 +82,34 @@ func (h *appHandler) setRevisionMetadata(appRev *v1beta1.ApplicationRevision) {
 // setRevisionWithRenderedResult will set the ApplicationRevision with the rendered result
 // it's ApplicationConfiguration and Component for now
 func (h *appHandler) setRevisionWithRenderedResult(appRev *v1beta1.ApplicationRevision, ac *v1alpha2.ApplicationConfiguration,
-	comps []*v1alpha2.Component) {
-	appRev.Spec.Components = ConvertComponent2RawRevision(comps)
+	comps []*v1alpha2.Component, policies []*unstructured.Unstructured) {
+	appRev.Spec.Components = ConvertComponents2RawRevisions(comps)
 	appRev.Spec.ApplicationConfiguration = util.Object2RawExtension(ac)
+
+	resources := []common.RawComponent{}
+	for _, c := range comps {
+		resources = append(resources, common.RawComponent{Raw: component2RawExtension(c)})
+	}
+	for _, acc := range ac.Spec.Components {
+		for _, tr := range acc.Traits {
+			resources = append(resources, common.RawComponent{Raw: util.Object2RawExtension(tr.Trait)})
+		}
+	}
+	for _, policy := range policies {
+		resources = append(resources, common.RawComponent{Raw: util.Object2RawExtension(policy)})
+	}
+
+	appRev.Spec.Resources = resources
+}
+
+func component2RawExtension(c *v1alpha2.Component) runtime.RawExtension {
+	u, err := util.Object2Unstructured(c.Spec.Workload)
+	if err != nil {
+		panic(err)
+	}
+	u.SetName(c.Name)
+	u.SetNamespace(c.Namespace)
+	return util.Object2RawExtension(u)
 }
 
 // gatherRevisionSpec will gather all revision spec withouth metadata and rendered result.
```

Comment:
We will take workload/policy into account for revision spec and computing hash, right?

Reply:
Yes. Everything except Workflow will be counted in revision hash.